src/hippopt/base/optimization_solver.py

```diff
@@ -0,0 +1,48 @@
+import abc
+import dataclasses
+from typing import List, Tuple, TypeVar
+
+import casadi as cs
+
+from hippopt.base.optimization_object import TOptimizationObject
+
+TOptimizationSolver = TypeVar("TOptimizationSolver", bound="OptimizationSolver")
+
+
+@dataclasses.dataclass
+class OptimizationSolver(abc.ABC):
+    @abc.abstractmethod
+    def generate_optimization_objects(
+        self, input_structure: TOptimizationObject | List[TOptimizationObject], **kwargs
+    ) -> TOptimizationObject | List[TOptimizationObject]:
```

Comment:
While it's intuitive returning a single `TOptimizationObject` in case only one is generated, in practice it might be better to always return a list, with just one element in this case. Otherwise, downstream users that want to write generic code need to add logic to handle the two cases. If they expect just a single object, the can always hardcode `out[0]`.